src/libraries/System.Web.HttpUtility/src/System/Web/Util/HttpEncoder.cs

```diff
@@ -480,6 +450,50 @@ internal static string UrlDecode(ReadOnlySpan<char> value, Encoding encoding)
             return expandedBytes;
         }
 
+        // Set of safe chars, from RFC 1738.4 minus '+'
+        private static readonly SearchValues<byte> s_urlSafeBytes = SearchValues.Create(
+            "!()*-.0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ_abcdefghijklmnopqrstuvwxyz"u8);
+        private static bool NeedsEncoding(ReadOnlySpan<byte> bytes, out int cUnsafe)
+        {
+            cUnsafe = 0;
+
+            int i = bytes.IndexOfAnyExcept(s_urlSafeBytes);
+            if (i < 0)
+            {
+                return false;
+            }
+
+            foreach (byte b in bytes.Slice(i))
+            {
+                if (!s_urlSafeBytes.Contains(b) && b != ' ')
+                {
+                    cUnsafe++;
+                }
+            }
+
+            return true;
+        }
+
+        internal static byte[] UrlEncode(string str, Encoding e)
```

Comment:
Would it make sense to split this method into a stack-alloc path and the fallback path w/ the byte-array?

Something like
```c#
internal static byte[] UrlEncode(string str, Encoding e)
{
    if (e.GetMaxByteCount(str.Length) <= StackallocThreshold)
    {
        Span<byte> byteSpan = stackalloc byte[StackallocThreshold];
        int encodedBytes = e.GetBytes(str, byteSpan);

        return UrlEncode(byteSpan.Slice(0, encodedBytes));
    }

    return Fallback(str, e);

    [MethodImpl(MethodImplOptions.NoInlining)]
    static byte[] Fallback(string str, Encoding e)
    {
        byte[] bytes = e.GetBytes(str);
        if (!NeedsEncoding(bytes, out int cUnsafe))
        {
            // return encoded byte[] if nothing to expand
            return bytes;
        }

        return UrlEncode(bytes, cUnsafe);
    }
}
```
so that the stack-alloc path becomes smaller and potentially inlinable?